Implement roll-forward policies for .NET Core SDK resolution.

peterhuene opened the pull request:
This PR implements additional roll-forward policies for resolving a .NET
Core SDK based upon new SDK settings in `global.json`.

Two new values are now supported in `global.json` in the `sdk` settings:

* `rollForward` - the roll-forward policy to use (see below).
* `allowPrerelease` - controls whether or not preview SDKS may be resolved
(defaults to `true` to allow resolving prerelease SDKs).  


The new roll-forward policies are:

* `patch` - If the requested SDK is installed, use it. Otherwise, use the
  latest installed patch level that matches the requested major, minor, and
  feature band.
* `feature` - If the requested major, minor, and feature band is installed, use
the latest patch level for the specified feature band.  Otherwise, roll-forward
to the next available feature band and use the latest patch level for that
feature band.
* `minor` - If the requested major, minor, and feature band is installed, use
the latest patch level for the specified feature band.  Otherwise, roll-forward
to the next available feature band available for the same major version and use
the latest patch level for that feature band.
* `major` - If the requested major, minor, and feature band is installed, use
the latest patch level for the specified feature band.  Otherwise, roll-forward
to the next available feature band available without restriction and use the
latest patch level for that feature band.
* `latestPatch` - Use the latest installed patch level that matches the
requested major, minor, and feature band.
* `latestFeature` - Use the latest installed patch level for the latest
installed feature band that matches the requested major and minor version.
* `latestMinor` - Use the latest installed patch level for the latest installed
feature band for the latest installed minor version that matches the requested
major version.
* `latestMajor` - Use the absolutely latest .NET Core SDK available.

For backwards compatibility, the policy when a `version` is not specified in
`global.json` will be `latestMajor`.  The default policy when a `version` is
specified but `rollForward` is not specified will be `patch`.

Commit:
Implement roll-forward policies for .NET Core SDK resolution.

This commit implements additional roll-forward policies for resolving a .NET
Core SDK based upon new SDK settings in `global.json`.

Two new values are now supported in `global.json` in the `sdk` settings:

* `rollForward` - the roll-forward policy to use (see below).
* `allowPrerelease` - controls whether or not preview SDKS may be resolved.

The new roll-forward policies are:

* `patch` - If the requested SDK is installed, use it. Otherwise, use the
  latest installed patch level that matches the requested major, minor, and
  feature band.
* `feature` - If the requested major, minor, and feature band is installed, use
the latest patch level for the specified feature band.  Otherwise, roll-forward
to the next available feature band and use the latest patch level for that
feature band.
* `minor` - If the requested major, minor, and feature band is installed, use
the latest patch level for the specified feature band.  Otherwise, roll-forward
to the next available feature band available for the same major version and use
the latest patch level for that feature band.
* `major` - If the requested major, minor, and feature band is installed, use
the latest patch level for the specified feature band.  Otherwise, roll-forward
to the next available feature band available without restriction and use the
latest patch level for that feature band.
* `latestPatch` - Use the latest installed patch level that matches the
requested major, minor, and feature band.
* `latestFeature` - Use the latest installed patch level for the latest
installed feature band that matches the requested major and minor version.
* `latestMinor` - Use the latest installed patch level for the latest installed
feature band for the latest installed minor version that matches the requested
major version.
* `latestMajor` - Use the absolutely latest .NET Core SDK available.

For backwards compatibility, the policy when a `version` is not specified in
`global.json` will be `latestMajor`.  The default policy when a `version` is
specified but `rollForward` is not specified will be `patch`.

peterhuene commented:
See [design document](https://github.com/dotnet/designs/pull/71) for more information on this feature.

peterhuene commented:
Investigating the windows failures (first guess would be path formatting in the tests).

peterhuene commented:
Nope, build break due to wchar_t on Windows.  Fixing.

nguerrera commented:
> The default policy when a version is
specified but rollForward is not specified will be patch.

I thought we agreed to change this to major.

peterhuene commented:
I wasn't sure if that was now captured in the design document, but you're right. It's easy to change and update the tests. I'll push that up and we can fix any language needed in the doc.

peterhuene commented:
We're going to discuss changing the default when just a version is specified from `patch` to `major` later today, as Kathleen has raised some concerns.  If we do decide to change the default, I'll push up that change at that time.  Until then, this can be reviewed with no change in the existing behavior.

johnbeisner commented on src/corehost/cli/fxr/fx_muxer.cpp:
```diff
@@ -969,8 +969,8 @@ int fx_muxer_t::handle_cli(
 
-    pal::string_t sdk_dotnet;
-    if (!sdk_resolver_t::resolve_sdk_dotnet_path(host_info.dotnet_root, &sdk_dotnet))
+    auto sdk_dotnet = sdk_resolver::from_nearest_global_file().resolve(host_info.dotnet_root);
```
Does this work when the SDK/SharedHost is installed and registered to a non-programfiles install location?

peterhuene replied:
This is no change in the current behavior.  This will use a SDK resolver based on the nearest global.json from CWD and use `host_info.dotnet_root` as the dotnet install root for resolving SDKs.  So if `host_info.dotnet_root` points at the non-program files location (either via environment or set from where `dotnet` is executing from), then we'll resolve SDKs from that install location.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -5,183 +5,240 @@
-    pal::readdir_onlydirectories(sdk_path, &versions);
-    fx_ver_t max_ver;
-    for (const auto& version : versions)
+sdk_roll_forward_policy sdk_resolver::to_policy(const pal::string_t& name)
```
This member needs to support case-insensitive names based on a discussion from the design document.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -192,79 +249,374 @@ bool sdk_resolver_t::resolve_sdk_dotnet_path(
-        if (global_cli_version.empty())
+    if (!sdk->second.is_object())
+    {
+        trace::error(_X("Expected a JSON object for the 'sdk' value in [%s]"), global_file_path.c_str());
```
I'm fine with using warnings as we do continue to just fallback as if the global.json didn't exist.  I'll get that fixed.

nguerrera replied:
Given that we won't be able to plumb warnings to VS without [https://github.com/dotnet/core-setup/issues/6972](https://github.com/dotnet/core-setup/issues/6972), I think it would be good to report this as an error. 

---
In reply to: [298392322](https://github.com/dotnet/core-setup/pull/6953#discussion_r298392322) [](ancestors = 298392322)

peterhuene replied:
I missed this feedback from a few days ago.  Given @vitek-karas preferred a warning in this context, but we can't currently report them as warnings from the VS resolver, should we change back to errors until #6972 is implemented?

vitek-karas replied:
Personally I'm not a big fan of writing to stderr without actually failing... but I get that in this case it might be helpful. The main reason I worry about this is things like msbuild - sometimes when msbuild runs a child process which writes to stderr it will consider the task a failure and fail the build (I think it depends on how the logging is setup for the child process).

One other consideration is if this is something we can change in the future. The code lives in `hostfxr` which is shared by all runtimes on the machine, so as such it is not exactly versioned. So even in .NET 5 we can't introduce breaking changes to it. In theory switching from stderr to stdout/warnings could be considered a breaking change.

I still think it would be better to keep these as warnings and tackle the issue of reporting warnings in VS.

peterhuene replied:
@nguerrera any strong opposition of merging this as-is and instead focusing on fixing #6972 to get the messages back?

peterhuene replied:
@vitek-karas this PR, as currently implemented, does include one error -> warning change, which is when global.json can't be parsed.  We used to error and ignore the file.  Now we warn about the parse failure and ignore the file with an additional warning saying the latest SDK will be used so users understand the file is being ignored.  Should we consider that breaking?

vitek-karas replied:
I think that one is much less worrisome, but technically it is. The reason why this is much easier to accept is that if something watches stderr and marks everything there as error, it will simply miss this one, but since the process still runs - probably not a big deal.

That said, thanks for pointing it out, I missed it. I think we should probably revert that one change back. It's just very late in the release cycle, unless we have a good reason to break, we should not. Even if I don't like it... 😄

peterhuene replied:
Alright, I will restore any errors we were previously emitting when handling global.json.  Any new diagnostics that aren't failures will remain warnings.

nguerrera replied:
> @nguerrera any strong opposition of merging this as-is and instead focusing on fixing #6972 to get the messages back?

No opposition at all :)

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -192,79 +249,374 @@ bool sdk_resolver_t::resolve_sdk_dotnet_path(
-        trace::verbose(_X("Searching SDK directory in [%s]"), dir.c_str());
-        pal::string_t current_sdk_path = dir;
-        append_path(&current_sdk_path, _X("sdk"));
+        trace::verbose(_X("UTF-8 BOM skipped while reading [%s]"), global_file_path.c_str());
```
I'll remove it.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.h:
```diff
@@ -3,18 +3,64 @@
+    sdk_resolver(fx_ver_t requested, sdk_roll_forward_policy policy, bool allow_prerelease);
+
+    pal::string_t const& global_file_path() const;
+    fx_ver_t const& requested_version() const;
```
I'll remove the unused members.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -5,183 +5,240 @@
+    auto requested = _requested.is_empty() ? pal::string_t{} : _requested.as_str();
+
+    trace::verbose(
+        _X("Resolving SDKs with version = '%s', roll-forward = '%s', allow-prerelease = %s"),
```
I'll use the JSON names for consistency.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.h:
```diff
@@ -3,18 +3,64 @@
+    bool resolve_sdk_path_and_version(const pal::string_t& dir, pal::string_t& sdk_path, fx_ver_t& resolved_version) const;
+
+    pal::string_t _global_file_path;
+    fx_ver_t _requested;
```
I'm fine with no prefix.  I'll remove them.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -5,183 +5,240 @@
-    pal::string_t* cli_sdk,
-    bool disallow_prerelease,
-    pal::string_t* global_json_path)
+const char* sdk_resolver::to_policy_name(sdk_roll_forward_policy policy)
```
I wasn't terribly worried about the optimization (might even get optimized out anyway) rather than making the logic pretty clear.  However, I agree that the redundant logic can be removed.  I'll fix that.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -192,79 +249,374 @@ bool sdk_resolver_t::resolve_sdk_dotnet_path(
+        case sdk_roll_forward_policy::latest_feature:
+            if (version_major != requested_major ||
+                version_minor != requested_minor ||
+                version_feature < requested_feature ||
```
I'll fix.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -192,79 +249,374 @@ bool sdk_resolver_t::resolve_sdk_dotnet_path(
+        return true;
+    }
+
+    // If there wasn't a requested version, then latest is best
```
I didn't want to implicitly rely on the default behavior in this member, but I can do so.  This was coded to be easy to understand, not necessarily the tightest code (especially given how corner cases always seem to creep into version handling).

peterhuene replied:
I'd prefer to leave the check in place so this member doesn't rely on the default and thus can guard against this internal state.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -192,79 +249,374 @@ bool sdk_resolver_t::resolve_sdk_dotnet_path(
+                return current_patch > previous_patch;
             }
+
+            return use_latest ? (current_feature > previous_feature) : (current_feature < previous_feature);
```
I think your suggestion is good.  I'll make that change.

peterhuene commented on src/corehost/cli/fxr/sdk_resolver.cpp:
```diff
@@ -192,79 +249,374 @@ bool sdk_resolver_t::resolve_sdk_dotnet_path(
+    return use_latest ? (current_major > previous_major) : (current_major < previous_major);
+}
+
+bool sdk_resolver::exact_match_allowed() const
```
That name works for me.  I'll change.

peterhuene commented on src/test/HostActivationTests/SDKLookup.cs:
```diff
@@ -555,6 +555,706 @@ public void SdkLookup_Must_Pick_The_Highest_Semantic_Version()
+            // Write a global.json that doesn't contain an object
+            WriteGlobalJson("true");
+
+             DotNet.Exec("help")
```
Will fix.

peterhuene commented on src/test/HostActivationTests/SDKLookup.cs:
```diff
@@ -555,6 +555,706 @@ public void SdkLookup_Must_Pick_The_Highest_Semantic_Version()
+                .And.HaveStdErrContaining(Path.Combine(_exeSelectedMessage, "9999.1.402", _dotnetSdkDllMessageTerminator));
+
+            // Write a global.json that has a non-string version
+            WriteGlobalJson("{ \"sdk\": { \"version\": 1 } }");
```
I'll see what I can do about theorizing this test.

peterhuene commented:
We've decided not to change the default behavior when a `version` is specified without `rollForward` policy in `global.json`.  The default will remain as `patch` for now.  Thus, these changes should not introduce any breaking changes for now.

Commit:
Make policy names case-insensitive in global.json.

This commit makes the policy names for the `sdk/rollForward` value in
`global.json` case insensitive.

It also removes the switch statements used in parsing / displaying the roll
forward policy names in favor of a constant array of names indexed by the enum
value.

Commit:
Address code review feedback.

* Removed unused members from `sdk_resolver`.
* Remove unnecessary logging relating to BOMs.
* Change to warnings instead of errors when failing to parse global.json (we
fallback to "use latest" behavior).
* Use JSON names in logging for values read in from global.json.
* Remove underscore prefix from member variables.
* Make `matches_policy` and `is_better_match` less redundant in the checks
performed.
* Rename `exact_match_allowed` to `exact_match_preferred`.
* Fix whitespace in test.

peterhuene commented:
Tests are failing because of 503 from nuget.org.  I still have a commit to theorize the requested test, so I'll just rerun the tests when that commit is pushed up.

Commit:
Theorize invalid global.json test.

This commit "theorizes" the invalid global.json test to eliminate duplicated
code.

nguerrera reviewed (approved):
:shipit:

vitek-karas reviewed (approved):
:shipit: assuming we have all the right approvals - even though in theory this should not be a breaking change, it's large enough to be treated similarly (especially since our test coverage is somewhat sparse and so it's hard to prove with reasonable confidence that this won't break anything).

Commit:
Keep JSON parse failure an error.

This commit turns the JSON parsing failure back to an error for backwards
compatibility.  The new diagnostic messages remain as warnings.

livarcocc commented:
This has been approved by Steve already. I gave him a heads up on this change a few weeks ago. Feel free to merge.

Also, just noticed the comment does not say what is the default for allowPrerelease. Should we state that in the description for the PR?

peterhuene commented:
I'll update the PR description to include that information.

peterhuene commented:
Oh right, I don't have access to merge 😄.

TheXenocide commented:
Is it possible to use these settings on items in the msbuild-sdks section? With NuGet we would use wildcard versions, but it doesn't seem to accept anything of the sort
